i6_models/parts/masked_norm.py

```diff
@@ -0,0 +1,127 @@
+    ):
+        """
+        :param num_features: feature dimension of the input.
+        :param eps: value added to the denominator for numerical stability.
+        :param momentum: Momentum for updating the running mean/variance statistics.
+            If set to none will use a cumulative average across all batches.
+        :param affine: whether to learn an affine transformation.
+        :param track_running_stats: whether to track running mean/variance statistics or
+            whether to just use single batch-based mean/variance.
+        :param device:
+        :param dtype:
+        """
+
+        super().__init__(
+            num_features,
+            eps=eps,
+            momentum=momentum,
+            affine=affine,
+            track_running_stats=track_running_stats,
+            device=device,
+            dtype=dtype,
+        )
+
+    def forward(self, inp: Tensor, lengths_or_mask: Tensor):
+        """
+        Applies batch norm to `inp`, masking away the padding given by `lengths_or_mask`.
+
+        :param inp: data to normalize, shape [B...,F,T]
+        :param lengths_or_mask: seq length tensor if shape [B...,],
+            or mask tensor if the shape is [B...,T].
+        """
+
+        self._check_input_dim(inp)
+
+        exponential_average_factor = 0.0
+        if self.training and self.track_running_stats and self.num_batches_tracked is not None:
+            self.num_batches_tracked += 1
+            if self.momentum is None:  # use cumulative moving average
+                exponential_average_factor = 1.0 / float(self.num_batches_tracked)
+            else:  # use exponential moving average
+                exponential_average_factor = self.momentum
+
+        assert inp.ndim - 3 < lengths_or_mask.ndim < inp.ndim, (
+            f"mask ndim ({lengths_or_mask.ndim}) should be between {inp.ndim - 3} and {inp.ndim}"
+        )
+        if lengths_or_mask.ndim == inp.ndim - 1:
+            mask = lengths_or_mask.to(dtype=torch.float, device=inp.device)
+        elif lengths_or_mask.ndim == inp.ndim - 2:
+            mask = _lengths_to_mask(lengths_or_mask, max_len=inp.shape[-1], dtype=inp.dtype)
+        else:
+            raise ValueError(
+                f"length tensor shape mismatch {lengths_or_mask.shape} wrt. input tensor shape {inp.shape}"
+            )
+        assert mask.ndim == 2
+
+        n = mask.sum()
+        if n == 0:
+            return inp
+
+        # we use the mask to calculate the mean
+        mask = mask / n
+        mask = mask.unsqueeze(-2)
+
+        reduce_dims = list(range(inp.ndim - 2)) + [inp.ndim - 1]
+        if not self.track_running_stats:
+            mean = (mask * inp).sum(reduce_dims)
+            var = ((mask * inp**2).sum(reduce_dims) - mean**2) * n / (n - 1)
+        elif self.training:
+            mean = (mask * inp).sum(reduce_dims)
+            # Var(X) = E[X^2] - E[X]^2
+            var = (mask * inp**2).sum(reduce_dims) - mean**2
```

Comment:
Why is there a distinction of having/not-having the factor `n / (n - 1)` in training vs (not training or not track_running_stats)? especially if we add the factor again to the running_var that we are tracking.

Reply:
Citing torch docs: 

> At train time in the forward pass, the variance is calculated via the biased estimator, equivalent to torch.var(input, unbiased=False). However, the value stored in the moving average of the variance is calculated via the unbiased estimator, equivalent to torch.var(input, unbiased=True).